typescript/sdk/src/ethers/retry-provider.ts

```diff
@@ -0,0 +1,81 @@
+// RetryProvider
+//
+// Mostly taken from the removed version that was in ethers.js
+// See: https://github.com/ethers-io/ethers.js/discussions/3006
+import { BaseProvider, JsonRpcProvider } from '@ethersproject/providers';
+import { ethers } from 'ethers';
+
+export type RetryOptions = {
+  // The wait interval in between
+  interval: number;
+
+  // Maximum number of times to rety
+  retryLimit: number;
+};
+
+type InFunction<T extends any[], U> = (...params: T) => Promise<U>;
+export function sleep(ms: number): Promise<void> {
+  return new Promise<void>((resolve) => setTimeout(resolve, ms));
+}
+
+// Retries an async function when it raises an exeption
+// if all the tries fail it raises the last thrown exeption
+export const retryAsync = async <InputArgs extends any[], ReturnType>(
+  inFunction: InFunction<InputArgs, ReturnType>,
+  tries: number,
+  params: InputArgs,
+  delay = 100,
+): Promise<ReturnType> => {
+  let saveError;
+  for (let i = 0; i < tries; i++) {
+    try {
+      // it awaits otherwise it'd always do all the retries
+      return await inFunction(...params);
+    } catch (error) {
+      await sleep(delay);
+      saveError = error;
+    }
+  }
+
+  throw saveError;
+};
+
+export class RetryProvider extends BaseProvider {
```

Comment:
Is this used?

Reply:
It's not because all our providers in the monorepo happened to be JsonRpcProviders, but other ones may not be